quantum/pointing_device_drivers.c

```diff
@@ -130,45 +402,57 @@ report_mouse_t cirque_pinnacle_get_report(report_mouse_t mouse_report) {
     // Scale coordinates to arbitrary X, Y resolution
     cirque_pinnacle_scale_data(&touchData, cirque_pinnacle_get_scale(), cirque_pinnacle_get_scale());
 
+    if (features.circular_scroll_enable) {
+        scroll_report  = circular_scroll(touchData);
+        mouse_report.v = scroll_report.v;
+        mouse_report.h = scroll_report.h;
+        if (scroll_report.suppress_touch) {
+            goto mouse_report_update;
+        }
+    }
     if (x && y && touchData.xValue && touchData.yValue) {
         report_x = (mouse_xy_report_t)(touchData.xValue - x);
         report_y = (mouse_xy_report_t)(touchData.yValue - y);
     }
-    x              = touchData.xValue;
-    y              = touchData.yValue;
-    mouse_report.x = report_x;
-    mouse_report.y = report_y;
-
-    if (touchData.touchDown != is_z_down) {
-        is_z_down = touchData.touchDown;
-        if (!touchData.zValue) {
-            if (timer_elapsed(mouse_timer) < CIRQUE_PINNACLE_TAPPING_TERM && mouse_timer != 0) {
-                mouse_report.buttons = pointing_device_handle_buttons(mouse_report.buttons, true, POINTING_DEVICE_BUTTON1);
-                pointing_device_set_report(mouse_report);
-                pointing_device_send();
-#    if TAP_CODE_DELAY > 0
-                wait_ms(TAP_CODE_DELAY);
-#    endif
-                mouse_report.buttons = pointing_device_handle_buttons(mouse_report.buttons, false, POINTING_DEVICE_BUTTON1);
-                pointing_device_set_report(mouse_report);
-                pointing_device_send();
+    x = touchData.xValue;
+    y = touchData.yValue;
+
+    if (features.cursor_glide_enable) {
+        if (touchData.touchDown) {
+            cursor_glide_update(report_x, report_y, touchData.zValue);
+        } else if (!glide_report.valid) {
+            glide_report = cursor_glide_start();
+            if (glide_report.valid) {
+                report_x = glide_report.dx;
+                report_y = glide_report.dy;
             }
         }
-        mouse_timer = timer_read();
     }
-    if (timer_elapsed(mouse_timer) > (CIRQUE_PINNACLE_TOUCH_DEBOUNCE)) {
-        mouse_timer = 0;
+
+mouse_report_update:
+    mouse_report.x = report_x;
+    mouse_report.y = report_y;
+
+    if (features.tap_enable) {
+        mouse_report = trackpad_tap(mouse_report, touchData);
     }
 
     return mouse_report;
 }
 
+uint16_t cirque_pinnacle_get_cpi(void) {
+    return (uint16_t)roundf((float)cirque_pinnacle_get_scale() * 25.4f / (float)CIRQUE_PINNACLE_DIAMETER_MM);
```

Comment:
I'm not a huge fan of using floats here, since AVR is still being used. 

If it helps, the lib8tion library is present already (used for rgb stuff mostly), and that may help.

Reply:
> However, it may be best to pull out the features that would work with other sensors (the inertial cursor code, perhaps), and move that to the pointing_devices.c file, while device specific features (such as the circular scroll) be moved into the cirque files.

The report structure currently doesn't support pulling these features to higher level code. Inertial cursor depends on knowing when a finger lifts, and circular scroll depends on absolute coordinates, both of which are not exposed by `report_mouse_t`. It might be possible to differentiate finger lift vs. stationary finger by the last velocity before `mouse_report.x` and `mouse_report.y` become 0, but that sounds pretty janky to me.
Seems like I should move these lower level to somewhere in _drivers/sensors/_ then, at least in the short term? I'm not sure about modifying the `get_report()` API for this one use case.

> I'm not a huge fan of using floats here, since AVR is still being used.
> If it helps, the lib8tion library is present already (used for rgb stuff mostly), and that may help.

For simple stuff like converting between inch and millimeter, sure I can get rid of the float usage.
I had tried earlier on to use lib8tion for the fancy trigonometry further in, but wasn't able to get it to behave correctly. Some inputs were too wide for lib8tion and had wrapped around before I could evaluate the accuracy.
Been running this as is on AVR; while it has trouble reporting at 1KHz on master side, it's still above 500Hz (tested by making inertial cursor reports run at 1ms interval and turning off friction coefficient).